src/main/kotlin/kotlinx/kover/appliers/KoverMergedApplier.kt

```diff
@@ -0,0 +1,255 @@
+
+        val notAppliedProjects =
+            projects.filter { it.extensions.findByType(KoverProjectConfig::class.java) == null }.map { it.path }
+        if (notAppliedProjects.isNotEmpty()) {
+            throw GradleException("Can't create Kover merge tasks: Kover plugin not applied in projects $notAppliedProjects")
+        }
+        projects.associateWith { it.extensions.getByType(KoverProjectConfig::class.java) }
+    }
+}
+
+
+private inline fun Project.mergedFilesProvider(
+    extensionByProject: Provider<Map<Project, KoverProjectConfig>>,
+    crossinline filterExtractor: (KoverProjectConfig) -> KoverSourceSetFilter
+): Provider<Map<String, ProjectFiles>> {
+    return provider {
+        extensionByProject.get()
+            .map { (project, extension) -> project.path to project.projectFiles(filterExtractor(extension), extension) }
+            .associate { it }
+    }
+}
+
+
+private fun filterProjects(filters: KoverProjectsFilter, allProjects: Iterable<Project>): List<Project> {
+    if (filters.includes.isEmpty()) {
+        return allProjects.toList()
+    }
+
+    val projectsByPath = allProjects.associateBy { p -> p.path }
+    val pathsByName = allProjects.associate { it.name to mutableListOf<String>() }
+    allProjects.forEach { pathsByName.getValue(it.name) += it.path }
+
+    return filters.includes.map {
+        if (it.startsWith(':')) {
+            projectsByPath[it]
+                ?: throw GradleException("Kover configuring error: not found project '$it' for merged tasks")
+        } else {
+            val paths = pathsByName[it]
+                ?: throw GradleException("Kover configuring error: not found project '$it' for merged tasks")
+            if (paths.size > 1) {
+                throw GradleException("Kover configuring error: ambiguous name of the project '$it' for merged tasks: suitable projects with paths $paths. Consider using fully-qualified name starting with ':'")
+            }
+            projectsByPath[paths[0]]!!
+        }
+    }
+}
+
+
+private fun Project.engineProvider(extensionByProject: Provider<Map<Project, KoverProjectConfig>>): Provider<EngineDetails> {
+    val archiveOperations: ArchiveOperations = this.serviceOf()
+    // configuration already created in all projects at this moment because merge tasks creates in afterEvaluate step
+    val config = configurations.getByName(CONFIGURATION_NAME)
+    // the plugin is always applied to the containing project
+    val containerEngine = extensions.getByType(KoverProjectConfig::class.java).engine
+
+    val containerPath = path
+    return provider {
+        val variants = extensionByProject.get().map { ComparableEngineVariant(it.value.engine.get()) }.toSet()
+        // check same engine variants are used
+        if (variants.size > 1) {
+            throw GradleException("Can't create Kover merge tasks: different coverage engines are used in projects")
```

Comment:
For here and all future exceptions you are going to throw: it is really important from both a UX standpoint of view and from our own good (e.g. to debug user or our own problems) to provide as much contextual information as possible: what are these versions and, if possible, what projects do carry these versions.

Taking lack of such practice to the extreme, there is nothing worse than getting a compiler BE error on a 100k+ lines project with the exception "Assertion failed: false" :)